### DISCUSSION TITLE
TuringCourses/major_module/embedded_system/

### DISCUSSION THREAD

**Discussion by giscus** [ZJU-Turing/TuringCourses#68] [General]
# TuringCourses/major_module/embedded_system/

图灵班学习指南

https://zju-turing.github.io/TuringCourses/major_module/embedded_system/

**Comment by ZhouTimeMachine** [comment]
特别澄清一下，靠一个人的努力是很难在 wzh 的课堂活下去的，实际上是大家结成统一战线，一致对外的结果。

    让我们来简单介绍一下 wzh 老师的上课风格吧。选 wzh 课的人本来就很少，都是被挤过来的，只有可怜的 30 几个人，中间受不了他的迫害又走了一批，貌似最后就剩了 10 几个人吧，以至于我们期末考的时候考场人少的像开组会一样。再加上 wzh 的课堂内容忽略不计，很多人都翘课不来听，后面基本每次课上只有 7-8 个人的样子。

    首先，因为是早八的课， wzh 老师意识到大三的学生早八上课还是太残忍了，于是把上课时间改到了 8:30，导致当时我们普遍认为 wzh 老师还是一位非常和蔼可亲的老师。结果我们发现 wzh 上课主要就是念一下 PPT，再加上很多 PPT 都是描述性的内容和大段的代码，以至于他会直接跳过这部分，一开始几节课还可以将 60min 左右的课程内容，到了后来甚至出现只念 10-20 min，甚至干脆不上课的现象，主要原因是 PPT 都念完了，或者“我们的课程进度比隔壁班快了，所以我们停一次课，大家做实验吧”，相对于时不时吹水的 wk 老师， wzh 老师的课除了干货也很难有别的东西了。之后 10 点开始上实验课，实验课也是念 10 min 的 PPT，然后自己做实验。所以我们的上课模式最后就变成了，9 点左右来了 7-8 个人， wzh 念 10 min PPT，然后我们自己做实验，等 10 点左右上实验课，wzh 再念 10 min PPT，我们继续做实验，所以做好自学的准备吧。

    不过和他的实验相比，上课这些问题实在是不值一提，没错，前方高能，做好准备。我们的实验内容总体是和 wk 班一样，但是顺序和实验内容略有不同，但是 wk 班提供了详细的实验文档，包括详细的实验文档和实验教程，而我们只有一份寡淡的 PPT，上面只有密密麻麻的实验要求，甚至连实验要求都不说清楚。

    我们举几个例子：lab1 是要求配置 firefly-rk3568 的开发板，尝试烧板子、配置各种软件。老师，不会怎么办？PPT给了官网的网址，自己查！没错，他所有的实验指导一言蔽之就是 —— 自己查。更离谱的是，lab1 的一个实验要求是逐行注释 rk3568 里面 ubuntu 系统的开机代码，开机代码有多少行呢？区区 1600 行吧，我注释了整整 3 天，最后 lab1 的报告写了 99 页，其中 60 多页是开机代码注释。另一个同学的导师是这么评价这个离谱的作业的——“艹”。

    后面的 STM 编程也没有任何的教程，还好 wk 班的同学提供了环境配置和编程的教程，救了我们全班的狗命。离谱的事情还没有结束，STM 板子和 PC 机之间通信是依靠一块串口板的，但是 wzh 没发，一开始我用 rk3568 的串口板顶替了一下啊，然后串口输出非常奇怪，以至于我根本无法确定到底是 PC、STM、软件、串口中哪一部分的问题，于是又 debug 了一天一夜！最后我向 jjgg 借了 wk 班的串口板，一插，好了，通了。于是我去向 wzh 要板子，才知道，他一直以为 rk3568 的串口板是可以替代 STM32 的串口板的。然后我才知道，他和他的团队仅仅拍脑子设计了实验，他们根本没有自己验证一下实验的正确性。一个星期后，发了新的板子，于是我们的实验都拖了一个星期。

    最神奇的是用 rk3568 的 GPIO 来做外设 LED 矩阵驱动的 lab7 。第一个小部分是用一个库来驱动 rk3568 的 GPIO， 但是很尴尬的什么呢，往年使用树莓派做 GPIO 驱动的，这个库是可以用在树莓派上的，但是我们的是 rk3568，根本用不了这个东西，最后和助教交涉了一下，才把这部分要求改掉！之后是做 GPIO 的内核驱动，wzh 友好的给了我们前几届（经考证，不晚于 2016 年）学长的代码，那岂不是只要编译运行就可以了！然后就被当头泼了冷水，因为实验平台改了，这个代码根本跑不起来……改了 3 天，然后下板子运行，没有任何反应，人麻了。更离谱的是，rk3568 的引脚是需要配套的线才可以连接，我们发的杜邦线比 GPIO 引脚大一圈，根本接触不良，没法导电，于是我们去找老师反映。老师很迷茫：这个实验助教做过的呀？问助教：这个实验我没做过…… wzh 说：接触不良的话，你找一个同学帮你按住杜邦线，用点力，让他接触良好，应该可以解决这个问题吧？拜托，五根线呀，那个方寸之地都摆不下五只手！

    于是想起了一个古老的笑话：一天大领导过来的时候，一个坐轮椅的同志没有站起来鞠躬，他的领导批评他：领导来了你为什么不站起来鞠躬？小同志说，我腿断了。领导说：这位小同志，我知道你有困难，但是苦难是可以想办法克服的嘛！你虽然腿断了，但是可以努力一下，自己站起来的嘛！所以，虽然杜邦线大了一圈接触不良，你可以努力一下找 3 个同学帮你按一下线，克服一下的嘛。

    最后经过我们的据理力争，wzh 答应给我们买 rk3568 的专用线，让我们下周一来拿，最后我们下周五拿到了线，至此两个星期过去了。然后试了各种方法驱动，STM32 法、shell 法、gpio 文件法、内核 gpio 的 api 法，茴香豆的回字也就 4 种写法呀！但是 LED 矩阵都没有反应。gpio 波形测了一下没问题，线没问题，LED 矩阵看起来也没问题，到底问题出在哪？中间我们怎么逼迫助教和 wzh 改实验的部分就不展开了，反正他们垂死挣扎，坚决不改实验，哪怕他们没人做过，宁可去问厂家怎么回事。第八周的周五，我们要去造反的时候，意外的换了一块 LED 板子，然后它亮了，然后我们才知道，虽然 LED 写了 VCC 要求 5V，但实际上它只能接受 3V 左右，之前的 LED 因为接了 5V 都被烧坏了，至此，三周过去了。

    反正 wzh 就是拍着脑子想实验，他只负责想当然的设计实验内容，根本不在乎实验做不做得出来；而助教呢，根本没做过实验，也没有提供过实验教程，我们全靠 wk 班的教程苟命，找他验收的时候我们说什么，他就听什么，到了下半个学期，这位助教出差去了，临时换了一位研究生学长来验收，那就更什么都不知道了。他们根本就不尊重我们的劳动，也不在乎我们的付出和时间。如果你不小心选上了 wzh 的课，不要把他当老师，他是你的甲方，是个甩手掌柜，只管布置任务，还不付你工资。

    最后是给分，你以为这么痛苦之后老师会因为心怀愧疚或者体谅给个高分吗？并不。现在离作业截至过去 5 天了，还没有出分数，平时分只有 3-4 次报告作业出分了，结果一看：我写了 21 页报告、做完所有 bonus、第一个验收的 lab2 给了 95 分，写了 14 页报告、第一个验收、做完所有 bonus 的 lab8 给了 90 分......其他人比我还有糟糕。我们接受不了这个结果，去和助教反应，结果第二天所有分数变成不公布了。所以 wzh 评分：bonus 少做扣分，验收越晚扣分，篇幅不够长扣分，交了报告直接扣 5 分。

    所以，不要选 wzh！不要选 wzh！不要选 wzh！补选 wk 一般还是补选的上的。如果实在时乖命蹇选上了 wzh 的课，或者说等看到这篇文章你来不及退课了，那就送大家 16 字方针，“独立自主，自力更生，统一战线，斗争到底”，反正好好发扬你们的斗争精神，分数、权益和幸福都是自己争取来的。

——转自 yygg

**Comment by NeilAlfred** [comment]
不论任何人，出于任何目的，我都不建议ta去上wzh老师的嵌入式系统。这个老师人本身就有问题。

**Comment by fishedCat** [comment]
21级应该只有我一个冤种选了。先说一下体验吧：
**我不推荐任何人选择这门课**
**我不推荐任何人选择这门课**
**我不推荐任何人选择这门课**
今年只有王总辉开设了这门课，分数分布还是和去年的一样。

### 上课体验
一共13次课程。分为理论和实验，是连在一起上的，先是两节理论，再是两节实验。
理论课纯念PPT，实验课纯念PPT。那能不能自己看PPT呢？**完全可以**
而且随着学期的推进，大家来的越来越晚，上课的时间也越来越晚，还有一次因为只有两个同学到了，所以老师就没有上课，我和另一个同学就在那里坐了一个上午。~~太高兴了~~
关于课堂表现，我们课上讲的是回答问题，他每节课都会提一些问题，有人回答的话会去他那里登记一下，之后他会综合给出表现分
对了，还要补充的是，这门课完全没有回放。

### 实验体验
**课8个小实验，一个3分，最少也要3天，最多的整整一周，对了，是净时间，不算上课之类的。
每个实验都有自己的雷点，硬件设备的配置是一个，根本不知道怎么才能配好，而且网上的资料很少，基本没有参考，他的PPT更是稀烂，要求真是纯要求，一点“废话”都没有。反正一个三分，实在做不出来的放了
ubuntu的开机注释确实是个大雷，1300多行，很多都找不到参考，看不懂它什么意思，不过现在有了GPT，可以丢给GPT去翻译一下
在lab4的时候他叫一组同学去讲解一下实验，之后每一组都要上去讲解一个实验，也算是个不错的尝试，但是到期末也就只有那一组讲解了实验……

大程是要求结合深度学习，使用实验板的摄像头，做一个项目，摄像头确实有很问题，一打开就是强光，晃得眼都快瞎了。开上一会就会很烫，建议现在电脑上跑一下程序，调好了再放到实验板上，用完就赶紧关上。

还要吐槽一下实验的时间安排，应该是两周一个的，但是从第一个就成了3周一个，那么当然的，期末就要爆炸了

### 期末考试体验
期末的题型是选择题，简答题，程序设计题
写得快的一个小时就结束了，我一个半小时交卷，是倒数第二个走的。
选择题主要是PPT上的一些知识点，没有超纲的，PPT可以找到所有的答案。
简答题的话不大记得了，有一题是问 bootloader的两个阶段主要干了什么。PPT上都有的。
程序设计的话第一题是设计一个LED的驱动程序，但是我不太明白它题目的意思，是写什么样的代码也没搞懂，不误导了。第二题是个智能家居系统的设计，传感器，嵌入式计算机，服务器搭一个能和手机通信的智能家居，然后二三小问加一个盗窃（门窗）检测传感器。
感觉考试还可以，不算多难，和他的介绍“都是嵌入式的常识”出入不大

### 总结
这门课的问题主要还是实验，实验的参考资料太少，难度不小，硬件调试不出来也没有办法，花的时间太多，相当的困难。我本来是希望在大三修完所有的课程，为了凑模块课的学分才选了嵌入式，但是其他课也不少，嵌入式是这个学期花的时间最多的一门课了。~~真的很多！！！~~
嵌入式的课程讲的东西还是不少的，如果真的想学嵌入式可以只看他的PPT，很快就能看完，和上课差不多。

**我不推荐任何人选择这门课！！！**
~~除非你和我一样，承压能力极强~~
